interp/src/flatten/structures/environment/wave.rs

```diff
@@ -0,0 +1,220 @@
+        };
+        let mut writer = open_fst(file_path, &info)?;
+        let port_map = declare_signals(&mut writer, env)?;
+        let writer = writer.finish()?;
+        Ok(Self {
+            writer: Some(writer),
+            port_map,
+        })
+    }
+
+    pub fn write_values(&mut self, time: u64, values: &PortMap) -> Result<()> {
+        let writer = self.writer.as_mut().unwrap();
+        writer.time_change(time)?;
+        for (port_id, maybe_signal_id) in self.port_map.iter() {
+            if let Some(signal_id) = maybe_signal_id {
+                match values[port_id].val() {
+                    None => {
+                        writer.signal_change(*signal_id, "x".as_bytes())?;
+                    }
+                    Some(value) => {
+                        writer.signal_change(
+                            *signal_id,
+                            value.to_bit_str().as_bytes(),
+                        )?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Drop for WaveWriter {
+    fn drop(&mut self) {
+        let writer = std::mem::take(&mut self.writer);
+        if let Some(writer) = writer {
+            writer.finish().unwrap();
+        }
+    }
+}
+
+enum Todo {
+    /// instance name, instance id
+    OpenScope(String, GlobalCellIdx),
+    CloseScope,
+}
+
+type PortToSignalMap = AuxillaryMap<GlobalPortIdx, Option<FstSignalId>>;
+
+fn declare_signals<
+    C: AsRef<Context> + Clone,
+    W: std::io::Write + std::io::Seek,
+>(
+    out: &mut FstHeaderWriter<W>,
+    env: &Environment<C>,
+) -> Result<PortToSignalMap> {
+    let mut port_map: PortToSignalMap = PortToSignalMap::new();
+    let root_idx = GlobalCellIdx::new(0);
+    let root_comp = &env.ctx.as_ref().secondary[env.ctx().entry_point];
+    let root_name = env.ctx().lookup_name(root_comp.name).clone();
+    let mut todo = vec![Todo::OpenScope(root_name, root_idx)];
+    while let Some(component) = todo.pop() {
+        match component {
+            Todo::OpenScope(name, id) => {
+                todo.push(Todo::CloseScope);
+                declare_component(
+                    out,
+                    env,
+                    &mut todo,
+                    &mut port_map,
+                    id,
+                    name,
+                )?;
+            }
+            Todo::CloseScope => {
+                out.up_scope()?;
+            }
+        }
+    }
+    Ok(port_map)
+}
+
+fn declare_component<
+    C: AsRef<Context> + Clone,
+    W: std::io::Write + std::io::Seek,
+>(
+    out: &mut FstHeaderWriter<W>,
+    env: &Environment<C>,
+    todo: &mut Vec<Todo>,
+    port_map: &mut PortToSignalMap,
+    component_cell_idx: GlobalCellIdx,
+    instance_name: String,
+) -> Result<()> {
+    let instance = env.cells[component_cell_idx].as_comp().unwrap();
+    let component = &env.ctx.as_ref().secondary[instance.comp_id];
+
+    let component_name = &env.ctx().lookup_name(component.name);
```

Comment:
remove the leading `&` since `lookup_name` returns a reference and I don't think
you need a `&&String`